org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ChangeSignatureHandler.java

```diff
@@ -0,0 +1,201 @@
+					logFatalError(status);
+					return null;
+				}
+			}
+			List<ParameterInfo> parameterInfos = processor.getParameterInfos();
+			List<ParameterInfo> newParameterInfos = new ArrayList<>();
+			for (MethodParameter param : parameters) {
+				if (param.originalIndex != ParameterInfo.INDEX_FOR_ADDED && param.originalIndex < parameterInfos.size()) {
+					ParameterInfo info = parameterInfos.get(param.originalIndex);
+					info.setNewTypeName(param.type);
+					info.setNewName(param.name);
+					newParameterInfos.add(info);
+				} else {
+					newParameterInfos.add(ParameterInfo.createInfoForAddedParameter(param.type, param.name, param.defaultValue));
+				}
+			}
+			parameterInfos.clear();
+			parameterInfos.addAll(newParameterInfos);
+			List<ExceptionInfo> exceptionInfos = processor.getExceptionInfos();
+			List<ExceptionInfo> newExceptionInfos = new ArrayList<>();
+			for (MethodException exception : exceptions) {
+				if (exception.typeHandleIdentifier != null) {
+					IJavaElement element = JavaCore.create(exception.typeHandleIdentifier);
+					if (element instanceof IType type) {
+						newExceptionInfos.add(ExceptionInfo.createInfoForAddedException(type));
+					}
+				} else {
+					IType type = null;
+					if (exception.type.equals("Exception")) {
+						// special handling for java.lang.Exception
+						type = cu.getJavaProject().findType("java.lang.Exception");
+					}
+					if (type == null) {
+						// find possible types with the fully qualified name in the project
+						type = cu.getJavaProject().findType(exception.type);
+					}
+					if (type == null) {
+						// find possible match types in existing import declarations
+						for (IImportDeclaration importDeclaration : cu.getImports()) {
+							String importName = importDeclaration.getElementName();
+							int dotIndex = importName.lastIndexOf(".");
+							if (dotIndex != -1 && dotIndex < importName.length() - 1) {
+								String typeName = importName.substring(dotIndex + 1);
+								if (typeName.equals(exception.type)) {
+									type = cu.getJavaProject().findType(importName);
+									break;
+								}
+							}
+						}
+					}
+					if (type == null) {
+						// exception.type is not FQN, or it should be found via IJavaProject.findType()
+						SearchEngine engine = new SearchEngine();
+						IJavaSearchScope scope = SearchEngine.createJavaSearchScope(new IJavaProject[] { cu.getJavaProject() }, true);
+						int qualIndex = exception.type.lastIndexOf('.');
+						if (qualIndex == -1) {
+							List<IType> foundTypes = new ArrayList<>();
+							engine.searchAllTypeNames(null, SearchPattern.R_FULL_MATCH, exception.type.toCharArray(), SearchPattern.R_FULL_MATCH, IJavaSearchConstants.TYPE, scope, new TypeNameMatchRequestor() {
+								@Override
+								public void acceptTypeNameMatch(TypeNameMatch match) {
+									IType type = match.getType();
+									try {
+										if (type.exists() && JdtFlags.isPublic(type)) {
+											foundTypes.add(type);
```

Comment:
Might not be correct here for example:

1. The type is package visible and in the same package of the current compilation unit.
2. The type is a static inner class: `public class A { static class B }`.

Most users may not meet the problem since above cases are not commonly used. It would be good if you can address them (maybe we already have similar logic somewhere), or at least file an issue to track the potential bugs here.

Reply:
Maybe: org.eclipse.jdt.internal.core.search.IRestrictedAccessTypeRequestor

Reply:
It looks like `IRestrictedAccessTypeRequestor` won't provide access information for types, let's track this in https://github.com/eclipse/eclipse.jdt.ls/issues/2541